core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,255 @@
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstream.onError(new IllegalArgumentException("non-positive request signals are illegal"));
+                return;
+            }
+            if (cancelled || state == State.DONE) {
+                return;
+            }
+            for (;;) {
+                final long oldRequested = requested;
+                final long newRequested = LongMath.saturatedAdd(oldRequested, n);
+                if (requestedUpdater.compareAndSet(this, oldRequested, newRequested)) {
+                    if (oldRequested > 0) {
+                        return;
+                    }
+                    break;
+                }
+            }
+
+            switch (state) {
+                case SENDING_HEAD: {
+                    setState(State.SENDING_HEAD, State.SENDING_BODY);
+                    if (head != null) {
+                        downstream.onNext(head);
+                        requestedUpdater.decrementAndGet(this);
+                    }
+                    publisher.subscribe(this);
+                    return;
+                }
+                case SENDING_BODY: {
+                    requestUpstream(upstream);
+                    return;
+                }
+                case SENDING_TAIL: {
+                    sendTail();
+                    if (n > 1) {
+                        sendComplete();
+                    }
+                    return;
+                }
+                case SENDING_COMPLETE: {
+                    sendComplete();
+                    return;
+                }
+                case DONE: {
+                    upstream.cancel();
+                    return;
+                }
+            }
+        }
+
+        private void sendTail() {
+            setState(State.SENDING_TAIL, State.SENDING_COMPLETE);
+            if (tail != null) {
+                requestedUpdater.decrementAndGet(this);
+                downstream.onNext(tail);
+            } else {
+                sendComplete();
+            }
+        }
+
+        private void sendComplete() {
+            if (state == State.DONE) {
+                return;
+            }
+            setState(State.SENDING_COMPLETE, State.DONE);
+            downstream.onComplete();
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            for (;;) {
+                final long requested = this.requested;
+                if (requested == 0) {
+                    return;
+                }
+                if (state == State.SENDING_TAIL) {
+                    sendTail();
+                    if (requested > 1) {
+                        sendComplete();
+                    }
+                    return;
+                }
+                if (requestedUpdater.compareAndSet(this, requested, 0)) {
+                    if (needToPublishUpdater.get(this) > Long.MAX_VALUE - requested) {
```

Comment:
Question: Should we remove `needToPublish` and use `requested` only to trace the remaining elements to publish? I'm asking because we can see a lot of atomic operations. I would like to remove them if possible.

Reply:
I try to remove `needToPublish` but can't find solution 😢

When `onComplete`, we don't know how many items to publish more cause we set `requested -> 0` on `requestUpstream()`.

e.g
- createPublisher(5L) -> request(5) -> `requested: 5 -> 0` -> onNext(5) -> expectNoEvent()
- createPublisher(5L) -> request(10) -> `requested: 10 -> 0` -> onNext(5) -> expectComplete()

requested counter is `0` on above cases so we need to know `needToPublishMore` to send `onComplete()` or not 🤔